File: java/MergeSort.java
public class MergeSort
  {
    public static void mergeSort(int[] arr)
    {
      mergeSortHelper(arr, 0, arr.length - 1);
    }
    private static void mergeSortHelper(int[] arr, int left, int right)
    {
      if (left < right)
      {
        int mid = left + (right - left) / 2;
        mergeSortHelper(arr, left, mid);
        mergeSortHelper(arr, mid+1, right);
        merge(arr, left, mid, right);
      }
    }
    private static void merge(int[] arr, int left, int mid, int right)
    {
      int n1 = mid - left + 1;
      int n2 = right - mid;
      int[] leftArray = new int[n1];
      int[] rightArray = new int[n2];

      for(int i = 0; i < n1; ++i) {
        leftArray[i] = arr[left + i];
      }
      for(int j = 0; j < n2; ++j) {
        rightArray[j] = arr[mid + 1 + j];
      }
      int i = 0, j = 0, k = left;
      while (i < n1 && j < n2) {
        if(leftArray[i] <= rightArray[j]) {
          arr[k] = leftArray[i];
          i++;
        }
        else
        {
          arr[k] = rightArray[j];
          j++;
        }
        k++;
      }
      while(i < n1) {
        arr[k] = leftArray[i];
        i++;
        k++;
      }
      while(j < n2) {
        arr[k] = rightArray[j];
        j++;
        k++;
      }
    }
  }

File: python/merge_sort.py
def merge_sort(arr):
  "" sorts array using merge sort""
  if len(arr) <= 1:
    return arr
  mid = len(arr) // 2
  left_half = arr[:mid]
  right_half = arr[mid:]

  left_half = merge_sort(left_half)
  right_half = merge_sort(right_half)

  return merge(left_half, right_half)

def merge(left, right):
  merged_array = []
  left_index = 0
  right_index = 0

  while left_index < len(left) and right_index < len(right):
    if left[left_index] <= right[right_index]:
      merged_array.append(left[left_index])
      left_index += 1
    else:
      merged_array.append(right[right_index])
      right_index += 1
  merged_array += left[left_index:]
  merged_array += right[right_index:]
  return merged_array 

# example usage
my_array = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]
sorted_array = merge_sort(my_array)
print(sorted_array)
  
